Add Java subproject allowing tasks to be programmed in Java with Python wrappers

Mac-Coleman opened the pull request:
This pull request is to add support for tasks programmed in Java to be run on the sign.

Short summary:
* The Maven project at `c4_sign/java_c4sign` can be compiled into a Jar file containing all the classes for the tasks written in Java.
* The Maven project can be setup automatically on Unix-based systems by running `bin/setup_java_project.sh`
* The Maven project can be compiled on Unix-based systems by running `bin/compile_java_project.sh`
* The main Python program should start the Java server automatically on Unix-based systems...
* The new Python class `JavaTask` is a wrapper for instances of Java classes that extend `JavaTaskBase`.
* The Java class `JavaTaskController` has a static method `getActiveTasks()` that returns an `ArrayList<JavaTaskBase>` of all the tasks that can be run.
* The Python program is made aware of these Java tasks when setting up all the tasks in `ScreenManager.update_tasks` by starting the Py4J gateway server, calling `JavaTaskController.getActiveTasks()`, and creating a Python `JavaTask` for each Java class discovered.

I'm not sure yet how robust my solution for setting up the Maven project is for everyone's install of Python, and it probably doesn't work natively for most Windows users...

Mac-Coleman committed:
Updated gitignore for Maven/Java Projects

Mac-Coleman committed:
First test of Java building with Maven

Adds Py4j as a dependency.

Mac-Coleman committed:
Added commons-lang3 as a java dependency

Mac-Coleman committed:
Added Constants.java

Mac-Coleman committed:
Added prepare and teardown methods to java base class

Mac-Coleman committed:
Added a random java task

Mac-Coleman committed:
First working version of a java task

Mac-Coleman committed:
Fixed name of task!

Mac-Coleman committed:
Added --no-upload flag

Skips uploading of histogram files

Mac-Coleman committed:
Changed RainbowWaveJava to more accurately reflect the Python RainbowWave.

Mac-Coleman committed:
Added py4j to setup.py

This adds py4j.jar to your python installations! Somewhere around
share/py4j/py4j0.x.x.x.jar...

Mac-Coleman committed:
Added short bash script for compiling the java project

Mac-Coleman committed:
Added a basic setup script to move py4j to the maven repository!

Mac-Coleman committed:
Added new error messages for setup failure

Mac-Coleman committed:
Added command to start java server

Mac-Coleman committed:
Automatically start and wait for gateway server, add --disable-java flag

Mac-Coleman committed:
Added maven to install script

Mac-Coleman committed:
Added setup and compile steps to install.sh

itsactuallyluna9 commented on bin/setup_java_project.sh:
```diff
@@ -0,0 +1,47 @@
+
+jar_name=$(basename $py4j_jar_location)
+
+jar_version=$(echo $jar_name | grep -oP "(?<=py4j).+(?=\.jar)")
```
`grep -oP` errors on macOS because `-P` is not an option

looking at solutions, I have to install grep (with `brew install grep`), and then change this to `ggrep`.

Mac-Coleman replied:
Yeah, I am thinking at this point that I just need to turn these into Python scripts to attempt to solve our compatibility problem. I should probably have known there would be a lot of differences what these commands support depending on the platform.

What directory should I place the new Python scripts in?

Mac-Coleman replied:
I've just decided to put everything in a new directory called `tools/`, so now we have:
1. `tools/setup_java_project.py`
2. `tools/compile_java_project.py`
3. `tools/start_java_server.py`

itsactuallyluna9 commented on c4_sign/java_c4sign/src/test/java/com/cornellcollegecomputingclub/java_c4sign/AppTest.java:
```diff
@@ -0,0 +1,19 @@
+     * Rigorous Test :-)
+     */
+    @Test
+    public void shouldAnswerWithTrue() {
```
ah man I don't think these tests will pass.... \j

itsactuallyluna9 commented on install.sh:
```diff
@@ -1,10 +1,12 @@
 sudo -H /home/c4/c4-sign/venv/bin/python3 -m pip install -e '.[physical]'
 sudo cp -v ./bin/c4-sign.service /lib/systemd/system/
+./bin/setup_java_project.sh
+./bin/compile_java_project.sh
```
so we run compile java project here, which is fine, but what about on sign startup (like if people update their code or add new code)?

Mac-Coleman replied:
Do you mean in `bin/start_sign.sh`? I should have done that.

I think that when the project is in development on the developer's machine, I would like to keep compiling as a manual step, maybe.

Mac-Coleman replied:
I've added a compile step to both the `install.sh` script and to the `start_sign.sh` script! I think...

itsactuallyluna9 commented on c4_sign/base_task.py:
```diff
@@ -299,3 +301,28 @@ def draw(self, canvas: Canvas, delta_time: timedelta):
             return False
         return super().draw(canvas, delta_time)
+
+class JavaTask(ScreenTask):
```
so, right now, simulator shows this as just `JavaTask`, and doesn't actually pull the name (which, for the python ones, is derived from from the class's name). might be worth adding in a thing or something for that?

Mac-Coleman replied:
I've changed it to work base on the ScreenTask's title, now, at least for overriding the current task in the simulator and for sending the task's names to the simulator. Is there anywhere else that we'd need to change that? Should we be making a different field for this purpose than just the title of the ScreenTask?

itsactuallyluna9 reviewed (changes requested):
overall, looks mostly good! there's just a few things that have to be changed :)

(there's also the whole windows thing we gotta figure out, but y'know)

itsactuallyluna9 committed:
Merge branch 'main' into java

itsactuallyluna9 commented:
updated branch to main so we have pretty simulator :)

itsactuallyluna9 committed:
fix screen manager in run_gif

Mac-Coleman committed:
Translated compile_java_project.sh to python

Mac-Coleman committed:
Translated start_java_server.sh to python

Mac-Coleman committed:
Translated setup_java_project.sh to python

Mac-Coleman committed:
Make python main program run python java server, change simulator task naming and overriding to be title-based

Maybe we should have some sort of canonical name for each task?

Mac-Coleman committed:
Change install and start scripts to use python scripts for managing java project

Mac-Coleman committed:
Delete java-related shell scripts (replaced with Python)

Mac-Coleman committed:
Changed error message to reflect new python script

Mac-Coleman committed:
Used Py4J "Launch Gateway" method instead of custom subprocess.Popen command. Should be more reliable...

Mac-Coleman committed:
Remove tools/start_java_server.py

The user should never need to start the java server on their own...

itsactuallyluna9 committed:
fix paths

Mac-Coleman committed:
Merge branch 'java' of https://github.com/CornellCollegeComputingClub/c4-sign into java

Mac-Coleman committed:
Revert "Merge branch 'java' of https://github.com/CornellCollegeComputingClub/c4-sign into java"

This reverts commit 77b6dcfc4a44e6a7e9add92f66e6ef155fcb363a, reversing
changes made to 72ccaf031dfb82a23dd0e384d58189b1d8391bcc.

Mac-Coleman committed:
Reapply fix paths

Mac-Coleman committed:
Returned Status from Java

itsactuallyluna9 committed:
add really horrible memory stress test

Mac-Coleman committed:
Bump Py4J version, update Java setup + compilation scripts

Mac-Coleman committed:
Switch to BufferedImage + Add Lissajous task

BufferedImage allows easy use of Graphics2D and WritableRaster

Mac-Coleman committed:
Added rainbow colors to Lissajous to demonstrate Java BufferedImage uses

Mac-Coleman committed:
Fixed operating system compatibility in build scripts

Mac-Coleman committed:
Fixed classpath to add commons-lang3 to py4j server

Mac-Coleman committed:
Fixed usage of JavaTaskBase's getLcdText method in python JavaTask class

Mac-Coleman committed:
Fixed JavaTaskBase getLcdText default implementation to match Python get_lcd_text default implementation

Mac-Coleman committed:
Added task for testing panel synchronization

Fixed imports, timedelta

Mac-Coleman committed:
Forgot to return true

Mac-Coleman committed:
Attempt to remove Neopixel class

Mac-Coleman committed:
Forgot to save file :(

Mac-Coleman committed:
Fix types before turning to Color

Mac-Coleman committed:
Fix: slices are only for position

Mac-Coleman committed:
fixed lambda definition

Mac-Coleman committed:
Deleted Neopixel class

Mac-Coleman committed:
Added sleep implementation to MatrixScreen to match SimulatorScreen

Mac-Coleman committed:
Disable TimingTest

Mac-Coleman committed:
Disable JVMStressTest. I still need to implement ignored tasks in Java

Mac-Coleman committed:
Artificial Horizon Task

Mac-Coleman committed:
Fix colors in java task

Mac-Coleman committed:
Add option to specify starting task for faster testing

Mac-Coleman committed:
Attempt gamma correction for sign

Mac-Coleman committed:
Limit brightness during gamma correction

Mac-Coleman committed:
Make 16 colors in Rainbow Wave instead of twelve

Mac-Coleman committed:
Fix broken condition in init_screen

Mac-Coleman committed:
New Double Pendulum task to test basic Verlet integration physics

Mac-Coleman committed:
Fixed bug causing screen to not be updated before override with --starting-task

Mac-Coleman committed:
Changing tasks that use complex numbers to use the python complex type

i didn't know it existed

Mac-Coleman committed:
Fixed teardown methods for Java Tasks

Mac-Coleman committed:
Added linear motion constraint for Verlet Particles

I have no idea if this is a good implementation but it's the first one
that came to mind

Mac-Coleman committed:
Added new "Cornell Gears" task

Spinning gears with Cornell "C"

Mac-Coleman commented:
Writing a note to myself to remember what to continue thinking about before merging this in:

Currently, the `--gif` option struggles with `JavaTask`s because the script that does the conversion uses the python classname for each task, which is "JavaTask" for all the tasks written in Java. I think we might have to make it so that each task has a separate, unique, "canonical" name, and currently the classname and task title are not suitable. If adding a new field for this "canonical" name sounds good to you, I will try that.

As long as we're adding fields to these classes, I think it might also be useful to have a description field where task artists can explain what they made and what significance it holds to them. This field could be written into `docs/screen_tasks.md` and could serve as a useful introduction to the project and what kinds of things it can do. Does that sound good?

Mac-Coleman committed:
Finally updated the --gif option to use canonical names instead of class names

Mac-Coleman committed:
Update main sign image

Mac-Coleman committed:
Updated README to have more thorough information about Java tasks

Mac-Coleman committed:
Updated documentation to explain why to choose Python or Java

Mac-Coleman committed:
Updated to use canonical names where possible

Mac-Coleman committed:
Updated matrix to use perf_counter_ns instead of arrow to reduce jitter

Mac-Coleman committed:
Forgot to use same units for delay

Mac-Coleman committed:
Hitch-finding test

Mac-Coleman committed:
Revert "Hitch-finding test"

This reverts commit 25762f2d6d166a72479a267ea197e8577b6148cf.
I don't know why I thought this would work
